command/log_flags.go

```diff
@@ -0,0 +1,142 @@
+type logFlags struct {
+	flagLogLevel          string
+	flagLogFormat         string
+	flagLogFile           string
+	flagLogRotateBytes    string
+	flagLogRotateDuration string
+	flagLogRotateMaxFiles string
+}
+
+type provider = func(key string) (string, bool)
+
+type valuesProvider struct {
+	flagProvider   provider
+	envVarProvider provider
+}
+
+// addLogFlags will add the set of 'log' related flags to a flag set.
+func (f *FlagSet) addLogFlags(l *logFlags) {
+	if l == nil {
+		l = &logFlags{}
+	}
+
+	f.StringVar(&StringVar{
+		Name:       flagNameLogLevel,
+		Target:     &l.flagLogLevel,
+		Default:    notSetValue,
+		EnvVar:     EnvVaultLogLevel,
+		Completion: complete.PredictSet("trace", "debug", "info", "warn", "error"),
+		Usage: "Log verbosity level. Supported values (in order of detail) are " +
+			"\"trace\", \"debug\", \"info\", \"warn\", and \"error\".",
+	})
+
+	f.StringVar(&StringVar{
+		Name:       flagNameLogFormat,
+		Target:     &l.flagLogFormat,
+		Default:    notSetValue,
+		EnvVar:     EnvVaultLogFormat,
+		Completion: complete.PredictSet("standard", "json"),
+		Usage:      `Log format. Supported values are "standard" and "json".`,
+	})
+
+	f.StringVar(&StringVar{
+		Name:   flagNameLogFile,
+		Target: &l.flagLogFile,
+		EnvVar: EnvVaultLogFile,
+		Usage:  "Path to the log file that Vault should use for logging",
+	})
+
+	f.StringVar(&StringVar{
+		Name:   flagNameLogRotateBytes,
+		Target: &l.flagLogRotateBytes,
+		EnvVar: EnvVaultLogRotateBytes,
+		Usage: "Number of bytes that should be written to a log before it needs to be rotated. " +
+			"Unless specified, there is no limit to the number of bytes that can be written to a log file",
+	})
+
+	f.StringVar(&StringVar{
+		Name:   flagNameLogRotateDuration,
+		Target: &l.flagLogRotateDuration,
+		EnvVar: EnvVaultLogRotateDuration,
+		Usage: "The maximum duration a log should be written to before it needs to be rotated. " +
+			"Must be a duration value such as 30s",
+	})
+
+	f.StringVar(&StringVar{
+		Name:   flagNameLogRotateMaxFiles,
+		Target: &l.flagLogRotateMaxFiles,
+		EnvVar: EnvVaultLogRotateMaxFiles,
+		Usage:  "The maximum number of older log file archives to keep",
+	})
+}
+
+// getFlagValue will attempt to find the flag with the corresponding key and return the value
+// along with a bool representing whether of not the flag had been found/set.
+func getFlagValue(fs *FlagSets, key string) (string, bool) {
+	var result string
+	var isFlagSet bool
+
+	fs.Visit(func(f *flag.Flag) {
+		if f.Name == key {
+			result = f.Value.String()
+			isFlagSet = true
+		}
+	})
+
+	return result, isFlagSet
+}
+
+// getAggregatedConfigValue uses the provided keys to check CLI flags and environment variables for values that may be
+// used to override any specified configuration. If nothing can be found the 'fallback' (default) value will be provided.
+func (p *valuesProvider) getAggregatedConfigValue(flagKey, envVarKey, current, fallback string) string {
```

Comment:
Docs are a good shout, I had just left it initially thinking I was going OTT with comments comments everywhere (there's always a fear that the comments and code diverge, so was trying to balance) but you pointing it out defo means I should have! 

If it's OK with you (happy to discuss further if required) could we just leave the `valuesProvider` as the idea of it was to have an implementation agnostic way to get the values inside the func, so that in the tests I can do away with any real calls to `os..` or have to create a real flag set, just have some basic provider that can do it all. (see: `TestLogFlags_ValuesProvider`).

Reply:
Ah, okay, so this is for dependency injection. I _think_ the idiomatic way to achieve this would be to have an interface that could be implemented by some mock type, but I am by no means a Go expert and this is certainly not a blocker 😄. Just proposing the idea for your consideration.